### PR TITLE
When uploading, automatically reduce the size of receipt image

### PR DESCRIPTION
**Pull request by nkdengineer**
### Fixed Issues

$ https://github.com/Expensify/App/issues/44084
PROPOSAL: https://github.com/Expensify/App/issues/44084#issuecomment-2188022852

### Tests

1. Login
2. Submit a manual expense
3. Add receipt
4. Upload a image with size > 24MB
5. Verify that: It works normally 
- [x] Verify that no errors appear in the JS console

### QA Steps

1. Login
2. Submit a manual expense
3. Add receipt
4. Upload a image with size > 24MB
5. Verify that: It works normally 

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/4f195628-8153-40c3-93ea-9f25a60b91f9

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/4d04bd4f-a7b2-40cb-859a-66a44f33df9b

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/ba2bf337-a894-4e6a-b0ec-b14513a1b7de

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/408c14ef-9e28-438c-98ed-a27c6716bac0

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/68db0cf1-7bbd-4d5b-a923-c9a322b699d3

</details>

### CODE REVIEW

File: src/libs/fileDownload/FileUtils.ts [hunk 1]
```diff
@@ -285,6 +286,27 @@ function isHighResolutionImage(resolution: {width: number; height: number} | nul
...
+    ImageSize.getSize(file.uri ?? '').then(({width, height}) => {
+        let newWidth;
+        let newHeight;
+        if (width < height) {
+            newHeight = CONST.MAX_IMAGE_DIMENSION;
+            newWidth = (newHeight * width) / height;
+        } else {
+            newWidth = CONST.MAX_IMAGE_DIMENSION;
+            newHeight = (newWidth * height) / width;
+        }
```
Comment (luacmartins) [inline src/libs/fileDownload/FileUtils.ts:299]: We can simplify this a bit

```suggestion
        const scaleFactor = CONST.MAX_IMAGE_DIMENSION / (width < height ? height : width);
        const newWidth = width * scaleFactor;
        const newHeight = height * scaleFactor;
```

**Review by luacmartins — CHANGES_REQUESTED** [review]
I haven't tested but code LGTM. Left a small suggestion to simplify the resizing code

File: src/libs/fileDownload/FileUtils.ts (hunk 1 above)
Comment (nkdengineer) [reply src/libs/fileDownload/FileUtils.ts]: Thanks your suggestion

**Comment by luacmartins** [comment]
@shubham1206agra I think any image over 2MB should do it

**Review by luacmartins — DISMISSED** [review]

**Comment by luacmartins** [comment]
@shubham1206agra bump for review

File: src/libs/fileDownload/FileUtils.ts [hunk 2]
```diff
@@ -285,6 +286,21 @@ function isHighResolutionImage(resolution: {width: number; height: number} | nul
...
+    ImageSize.getSize(file.uri ?? '').then(({width, height}) => {
+        const scaleFactor = CONST.MAX_IMAGE_DIMENSION / (width < height ? height : width);
+        const newWidth = width * scaleFactor;
+        const newHeight = height * scaleFactor;
+
+        return {width: newWidth, height: newHeight};
+    });
+
+const resizeImageIfNeeded = (file: FileObject) => {
+    if (!file || !Str.isImage(file.name ?? '') || (file?.size ?? 0) <= CONST.API_ATTACHMENT_VALIDATIONS.MAX_SIZE) {
```
[the remark this answers is not in the document]
Comment (nkdengineer) [reply src/libs/fileDownload/FileUtils.ts:299]: @shubham1206agra you can download the large image [here](https://drive.google.com/file/d/1dcW-YtWnT4MjVuaF6ojj3zClRldQjxvc/view?usp=sharing)

**Comment by luacmartins** [comment]
I'm running an adhoc build since the attachment flow can be a bit weird on the simulator.

**Comment by nkdengineer** [comment]
Checking this now.

**Comment by nkdengineer** [comment]
> 
 Screen.Recording.2024-07-29.at.9.50.30.PM.mov 
@nkdengineer This is giving weird result here.


I can reproduce this, checking to find the reason.

**Comment by nkdengineer** [comment]
@shubham1206agra @luacmartins This bug only happens on Safari and I think this warning may be why. I'm finding a solution to fix this.

> Canvas area exceeds the maximum limit (width * height > 16777216).

Screenshot 2024-07-30 at 12 11 30

**Comment by nkdengineer** [comment]
@shubham1206agra What do you think if we only do this on safari mWeb?

**Comment by nkdengineer** [comment]
@shubham1206agra What do you think about this message

cc @luacmartins 

Screenshot 2024-07-30 at 16 05 03

**Comment by nkdengineer** [comment]
@shubham1206agra I'm investigating to find a fix for this case.

**Comment by nkdengineer** [comment]
@shubham1206agra The context is [here](https://pqina.nl/blog/canvas-area-exceeds-the-maximum-limit/). Because the canvas area limit on IOS Safari is `4096 * 4096` and `manipulateAsync` function uses the original width and height to create the canvas of the original image, this canvas becomes invalid and then the black image appears after resizing.

```
canvas.width = imageSource.naturalWidth;
canvas.height = imageSource.naturalHeight;
```

Screenshot 2024-07-30 at 22 32 53


1. If we want to fix this, we need to implement it in `expo-image-manipulator` lib to divide the image into small canvases, get image data from the small canvases, and combine them into image data of the original image before resizing.

2. Display a warning as this comment https://github.com/Expensify/App/pull/45448#issuecomment-2257856584



cc @luacmartins

**Comment by luacmartins** [comment]
Asked in [Slack](https://expensify.slack.com/archives/C04878MDF34/p1722357131043479) for someone familiar with `expo-image-manipulator` to comment here.

**Comment by BrtqKr** [comment]
@luacmartins @nkdengineer 
Ok, so the diagnosis regarding the issue was correct - the files exceeding the limit were corrupted (the format signatures were correct tho), but...

It looks like [enforcing the canvas size within the limits before performing the operations](https://github.com/software-mansion-labs/expensify-app-fork/commit/c167b92de09130f7f52c225477ced98224fd0ec5) (on the image manipulator level) seems to be fixing the problem. It's a bit weird, because with those assumptions the canvas is smaller than the image being processed (for a brief moment), but the end result should be fulfilling our expectations. 

On the other hand, what should not be fulfilling our expectations is the processing time - I'd suggest at least adding the indicator that the image is being processed. Uploading images of that size is probably not a common case, but this might be slightly confusing. 

I'll try to send a PR during the week-end and discuss it in the image-manipulator repository directly, but for now that should be sufficient. 


https://github.com/user-attachments/assets/fff9ac11-f6cc-484d-bdfd-18c5ee2b062a

**Comment by luacmartins** [comment]
Nice! Thanks for the investigation @BrtqKr! Looking forward to the PR!

**Comment by BrtqKr** [comment]
@nkdengineer if you could apply this, it would be amazing. Let me know if you find anything related to this patch. If this gets merged before I discuss it with expo and the library gets released, I'll clean this patch and bump the version in a separate PR.

**Comment by nkdengineer** [comment]
@BrtqKr I will add the patch in this PR soon.

**Comment by nkdengineer** [comment]
@shubham1206agra Updated the patch for the resizing solution, I tested and it works well. Please help to test again, thanks.

**Comment by nkdengineer** [comment]
@shubham1206agra I updated, I missed re-installing the lib before creating the patch.

**Review by shubham1206agra — APPROVED** [review]

**Review by luacmartins — APPROVED** [review]
LGTM

**Pull request merged**